### PR TITLE
Android tools 30.0.5

### LINKED ISSUE #11: Update to 30.0.4 (Android 11)
platform-tools-30.0.4 is available now
https://android.googlesource.com/platform/system/core

One of the important feature of adb of this version is `adb pair` which available for Android 11+

[Connect to a device over Wi-Fi (Android 11+)](https://developer.android.com/studio/command-line/adb)

### PR DESCRIPTION
**Pull request by sgn**
- Close #11 
- Built with gcc 9.3.0 with musl libc on Void Linux
- Smoke tested with Android 11

### CODE REVIEW

File: patches/core/0025-span-fallback-to-gsl-span-if-std-span-not-exists.patch [hunk 1]
```diff
@@ -0,0 +1,207 @@
+From 2cab7f70515a8fa3ca488c937e5540e02661835d Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?=C4=90o=C3=A0n=20Tr=E1=BA=A7n=20C=C3=B4ng=20Danh?=
+ <congdanhqx@gmail.com>
+Date: Sun, 1 Nov 2020 13:01:48 +0700
+Subject: [PATCH] span: fallback to gsl::span if std::span not exists
```
Comment (nmeum) [inline patches/core/0025-span-fallback-to-gsl-span-if-std-span-not-exists.patch:5]: Also a GCC incompatibility or why is this necessary?

File: patches/core/0026-adb-disable-mDNS.patch [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ }
+ 
+diff --git a/adb/socket_spec.cpp b/adb/socket_spec.cpp
+index 5cad70d09..f143b6c58 100644
+--- a/adb/socket_spec.cpp
++++ b/adb/socket_spec.cpp
+@@ -196,24 +196,6 @@ bool socket_spec_connect(unique_fd* fd, std::string_view address, int* port, std
+             fd->reset(network_loopback_client(port_value, SOCK_STREAM, error));
+         } else {
+ #if ADB_HOST
```
Comment (nmeum) [inline patches/core/0026-adb-disable-mDNS.patch:45]: Maybe just replace this with `#if 0` to trigger the `#else`?

File: patches/libbase/0006-less-lambda-less-problem.patch [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+-                              "%s%.*s", new_line, length, message);
+-    }
+-
+-    // This should never fail, if it does and we set size_written to 0, which will skip this line
+-    // and move to the next one.
+-    if (size_written < 0) {
+-      size_written = 0;
+-    }
+-    chunk_position += size_written;
+-  };
```
Comment (nmeum) [inline patches/libbase/0006-less-lambda-less-problem.patch:76]: What's wrong with this lambda expression? Does it not compile with g++? Looks fairly straightforward to me on first sight.

File: vendor/CMakeLists.txt [hunk 4]
```diff
@@ -26,6 +30,13 @@ if(EXISTS "${ANDROID_PATCH_DIR}/")
 	endforeach(v)
 endif()
 
+find_package(Protobuf REQUIRED)
```
Comment (nmeum) [inline vendor/CMakeLists.txt:33]: If protobuf is now required, please also add it to the `Dependencies` section in `README.md` :pray:

File: .gitmodules [hunk 5]
```diff
@@ -22,3 +22,18 @@
...
+	url = https://android.googlesource.com/platform/system/libziparchive
+[submodule "vendor/base"]
+	path = vendor/base
+	url = https://android.googlesource.com/platform/frameworks/base
+[submodule "vendor/native"]
+	path = vendor/native
+	url = https://android.googlesource.com/platform/frameworks/native
+[submodule "compat/gsl"]
+	path = compat/gsl
+	url = https://github.com/microsoft/GSL.git
```
Comment (nmeum) [inline .gitmodules:39]: Could you briefly explain why this is needed?

**Review by nmeum — COMMENTED** [review]
Thanks for working on this :tada: Below just some questions that came up and things I noticed while scrolling through the changes. Will hopefully get around to doing some quick tests and making a new release soonish. Thanks again!

File: patches/core/0025-span-fallback-to-gsl-span-if-std-span-not-exists.patch (hunk 1 above)
Comment (sgn) [reply patches/core/0025-span-fallback-to-gsl-span-if-std-span-not-exists.patch:5]: Yes, gcc 9.3.0 hasn't ship `std::span` yet, it's a C++20 feature. It should be fine with gcc 10.

File: patches/libbase/0006-less-lambda-less-problem.patch (hunk 3 above)
Comment (sgn) [reply patches/libbase/0006-less-lambda-less-problem.patch:76]: There are nothing wrong with this lambda expression, except that it trigger gcc internal error on my computer. I'm not sure if it's my computer that's lacking memory at the time of compiling and stack size on musl-libc is not enough for this lambda.

File: vendor/CMakeLists.txt (hunk 4 above)
Comment (sgn) [reply vendor/CMakeLists.txt:33]: I'll add it into `README.md`

File: .gitmodules (hunk 5 above)
Comment (sgn) [reply .gitmodules:39]: This library was added to provide fallback implementation for `std::span` in gcc 9 and prior.
Historically, `span` was implemented on Microsoft GSL first, then propose to C++ LWG later. (And it's removed from GSL after it's accepted to C++20)

Comment (sgn) [reply .gitmodules:39]: Since gsl::span was removed from current GSL, we can just vendor `gsl::span` in.

**Comment by hrap1919** [comment]
Hi,

I am wondering,  the reason of `#if 0` in patches/core/0026-adb-disable-mDNS.patch

It seems this will prevent 
`fd->reset(network_connect(hostname, port_value, SOCK_STREAM, 0, error));`
which is needed for adb connect.

Or I downloaded some old version of patches?

**Comment by sgn** [comment]
> I am wondering what is the reason of `#if 0` in patches/core/0026-adb-disable-mDNS.patch

It's there to prevent the block *below* it.

> 
> It seems this will prevent
> `fd->reset(network_connect(hostname, port_value, SOCK_STREAM, 0, error));`
> which is needed for adb connect.

No, that `#if 0` doesn't prevent that block of codes.

> 
> Or I downloaded some old version of patches?

You've seen the correct version.


---

Anyway, it looks like `30.0.5` is out, I'll look into rebasing for 30.0.5

**Comment by hrap1919** [comment]
BTW it causes
 
$ ./adb connect 192.168.1.1
* daemon not running; starting now at tcp:5037
* daemon started successfully
adbd does not support arbitrary tcp connections

**Comment by sgn** [comment]
@hrap1919 You're right.

Originally, I remove the block inside `#if ADB_HOST`, then @nmeum suggested to flip it, and I followed their advice mindlessly. Hence, it's jump into the `#else` block, which is incorrect.

File: patches/core/0026-adb-disable-mDNS.patch (hunk 2 above)
Comment (sgn) [reply patches/core/0026-adb-disable-mDNS.patch]: No, we don't want to jump into `#else` block. The whole block should be removed instead.

**Comment by nmeum** [comment]
> Originally, I remove the block inside `#if ADB_HOST`, then @nmeum suggested to flip it, and I followed their advice mindlessly. Hence, it's jump into the `#else` block, which is incorrect.

Oops, sorry. Just seemed the logical thing to do just by briefly skipping over the code :sweat_smile:

File: .gitmodules (hunk 5 above)
Comment (nmeum) [reply .gitmodules:39]: So with `gcc >= 10` this is not required?

**Review by nmeum — CHANGES_REQUESTED** [review]
The new version also seems to require `lz4`:

```
/home/soeren/src/android-tools/vendor/core/adb/compression_utils.h:26:10: fatal error: lz4frame.h: No such file or directory
   26 | #include <lz4frame.h>
      |          ^~~~~~~~~~~~
compilation terminated.
```

and `zstd`:

```
/home/soeren/src/android-tools/vendor/core/adb/compression_utils.h:27:10: fatal error: zstd.h: No such file or directory
   27 | #include <zstd.h>
      |          ^~~~~~~~
compilation terminated.
```

Please also add those two the `Dependency` section in `README.md`.

Otherwise compiles fine here :tada:

**Comment by sgn** [comment]
On 2020-11-12 10:31:25-0800, nmeum <notifications@github.com> wrote:
> @nmeum requested changes on this pull request.
> 
> The new version also seems to require `lz4`:
> 
> ```
> /home/soeren/src/android-tools/vendor/core/adb/compression_utils.h:26:10: fatal error: lz4frame.h: No such file or directory
>    26 | #include <lz4frame.h>
>       |          ^~~~~~~~~~~~
> compilation terminated.
> ```
> 
> and `zstd`:
> 
> ```
> /home/soeren/src/android-tools/vendor/core/adb/compression_utils.h:27:10: fatal error: zstd.h: No such file or directory
>    27 | #include <zstd.h>
>       |          ^~~~~~~~
> compilation terminated.
> ```
> 
> Please also add those two the `Dependency` section in `README.md`.

OK, my machine always has those headers :shrug:
I'll add them, now.

> 
> Otherwise compiles fine here :tada:

Nice!

File: .gitmodules (hunk 5 above)
Comment (sgn) [reply .gitmodules:39]: > So with `gcc >= 10` this is not required?

Yes, that's correct. `std::span` is avaiable with gcc 10 and clang 7+

**Comment by nmeum** [comment]
> OK, my machine always has those headers :shrug: I'll add them, now.

Thanks!

Would you mind if we wait with merging this until gcc 10 makes it into the void repositories (https://github.com/void-linux/void-packages/pull/24258)? That would allow us to drop the `gls::span`-related changes and the lambda expression from https://github.com/nmeum/android-tools/pull/12#discussion_r520057877 should *hopefully* also compile with gcc 10 thus making the required patches a bit simpler.

**Comment by sgn** [comment]
On 2020-11-22 22:58:54-0800, nmeum <notifications@github.com> wrote:
> > OK, my machine always has those headers :shrug: I'll add them, now.
> 
> Thanks!
> 
> Would you mind if we wait with merging this until gcc 10 makes it

Sure!
However, that PR wouldn't be merged into VoidLinux before linux 5.10.

We're waiting for Linux 5.10 to be ready to bump linux-kernel-headers
to 5.10 first. After it's settled down, we'll merge gcc 10 in.

> into the void repositories
> (https://github.com/void-linux/void-packages/pull/24258)? That would
> allow us to drop the `gls::span`-related changes and the lambda

The `std::span` fallback will be fixed with gcc 10, but I'm not sure
about the lambda expression. Let's me check its status this weekend.

> expression from
> https://github.com/nmeum/android-tools/pull/12#discussion_r520057877
> should *hopefully* also compile with gcc 10 thus making the required
> patches a bit simpler.

-- 
Danh

**Comment by nmeum** [comment]
Merged this but removed the lambda expression patch (compiles fine with gcc 10.X here) and the gls compat code (which isn't needed anymore with gcc >= 10.X). I also want to test this on a glibc-based distro and will then make a new release. Thanks for your work on this! :tada:
